### PR TITLE
python310Packages.cryptography: 39.0.1 -> 40.0.1

### PR DESCRIPTION
**Pull request by mweinelt**
https://cryptography.io/en/latest/changelog/#v40-0-1

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/python-modules/cryptography/default.nix [hunk 1]
```diff
@@ -28,26 +29,27 @@ let
...
     name = "${pname}-${version}";
-    hash = "sha256-0x+KIqJznDEyIUqVuYfIESKmHBWfzirPeX2R/cWlngc=";
+    hash = "sha256-gFfDTc2QWBWHBCycVH1dYlCsWQMVcRZfOBIau+njtDU=";
   };
 
   cargoRoot = "src/rust";
 
   nativeBuildInputs = lib.optionals (!isPyPy) [
     cffi
+    pkg-config
```
Comment (mweinelt) [inline pkgs/development/python-modules/cryptography/default.nix:52]: That should probably have been added unconditionally. Will send a follow-up PR.

Comment (mweinelt) [reply pkgs/development/python-modules/cryptography/default.nix:52]: Nope, builds without pkg-config, because it also doesn't use cffi.

**Comment by trofi** [comment]
Bisect says 5a5994d2adb7d55818fcffd8febcf8e83f45fb9e `python310Packages.cryptography: 39.0.1 -> 40.0.1` broke `python3Packages.pyopenssl` build in `staging`:

```
$ nix build --no-link -f. python3Packages.pyopenssl -L
...
python3.10-pyopenssl> installing
python3.10-pyopenssl> Executing pipInstallPhase
python3.10-pyopenssl> /build/pyOpenSSL-23.0.0/dist /build/pyOpenSSL-23.0.0
python3.10-pyopenssl> Processing ./pyOpenSSL-23.0.0-py3-none-any.whl
python3.10-pyopenssl> ERROR: Could not find a version that satisfies the requirement cryptography<40,>=38.0.0 (from pyopenssl) (from versions: none)
python3.10-pyopenssl> ERROR: No matching distribution found for cryptography<40,>=38.0.0
```

**Comment by mweinelt** [comment]
Fixed in #225138

**Pull request merged**